### PR TITLE
Add support for Piola mapping in 2D mixed ASMs

### PR DESCRIPTION
**Pull request by akva2**
This has now been tested and works within the expected limitations (in particular no boundary condition handling). Let's get it in before it gets lost in some dark git corner.

### CODE REVIEW

**Review by kmokstad — APPROVED** [review]
with some suggestions for (cosmetic) improvements.

File: src/ASM/BasisFunctionCache.h [hunk 1]
```diff
@@ -72,6 +72,9 @@ template<size_t Dim> class BasisFunctionCache
   //! \brief Returns whether or not a reduced quadrature is enabled.
   bool hasReduced() const { return !reducedQ->gpar[0].empty(); }
 
+  //! \brief Returns basis cache is for.
+  int getBasis() const { return basis; }
```
[the remark this answers is not in the document]
Comment (akva2) [reply src/ASM/BasisFunctionCache.h]: i generally prefer to keep as much as possible protected so i can easily control where it is mutated, but okay.

File: src/ASM/LR/ASMu2Dmx.C [hunk 2]
```diff
@@ -336,7 +337,10 @@ bool ASMu2Dmx::integrate (Integrand& integrand,
 
   for (std::unique_ptr<ASMu2D::BasisFunctionCache>& cache : myCache) {
     cache->setIntegrand(&integrand);
-    cache->init(use2ndDer ? 2 : 1);
+    if (cache->getBasis() == ASM::GEOMETRY_BASIS)
+      cache->init(piolaMapping || use2ndDer ? 2 : 1);
+    else
+      cache->init(use2ndDer ? 2 : 1);
```
[the remark this answers is not in the document]
Comment (akva2) [reply src/ASM/LR/ASMu2Dmx.C]: ok

File: src/ASM/ASMs2Dmx.C [hunk 3]
```diff
@@ -456,7 +457,10 @@ bool ASMs2Dmx::integrate (Integrand& integrand,
 
   for (std::unique_ptr<BasisFunctionCache>& cache : myCache) {
     cache->setIntegrand(&integrand);
-    cache->init(use2ndDer ? 2 : 1);
+    if (cache->getBasis() == ASM::GEOMETRY_BASIS)
+      cache->init(use2ndDer || piolaMapping ? 2 : 1);
+    else
+      cache->init(use2ndDer ? 2 : 1);
```
[the remark this answers is not in the document]
Comment (akva2) [reply src/ASM/ASMs2Dmx.C]: ok.

File: src/ASM/ASMs2D.h [hunk 4]
```diff
@@ -109,8 +109,7 @@ class ASMs2D : public ASMstruct, public ASM2D
 
     const ASMs2D& patch; //!< Reference to patch cache is for
 
-    int basis; //!< Basis to use
-    std::array<size_t,2> nel; //!< Number of elements in each direction
+    std::array<size_t,2> nel{0,0}; //!< Number of elements in each direction
```
[the remark this answers is not in the document]
Comment (akva2) [reply src/ASM/ASMs2D.h]: it's mainly defensive initialization, i have added it to the 3D class as well.

File: src/ASM/LR/ASMu2Dmx.C [hunk 5]
```diff
@@ -876,6 +876,76 @@ bool ASMu2Dmx::evalSolution (Matrix& sField, const Vector& locSol,
...
+                  [](const unsigned char in) { return in > 1; })) {
+    std::cerr << "*** ASMs2Dmx::evalSolution: Piola mapping requires "
+                 "a single field on each basis" << std::endl;
+    return false;
+  }
+
+  size_t len = std::accumulate(nb.begin(),nb.end(),0u);
+  if (len != locSol.size()) {
+    std::cerr << "*** ASMs2Dmx::evalSolution: Unexpected solution size"
+                 ", expected " << len << ", got " << locSol.size() << std::endl;
```
[the remark this answers is not in the document]
Comment (akva2) [reply src/ASM/LR/ASMu2Dmx.C]: ok

File: src/ASM/LR/ASMu2Dmx.C [hunk 6]
```diff
@@ -876,6 +876,76 @@ bool ASMu2Dmx::evalSolution (Matrix& sField, const Vector& locSol,
...
+                                  const RealArray* gpar, bool) const
+{
+  size_t nPoints = gpar[0].size();
+  if (nPoints != gpar[1].size())
+    return false;
+
+  if (std::any_of(nfx.begin(), nfx.end(),
+                  [](const unsigned char in) { return in > 1; })) {
+    std::cerr << "*** ASMs2Dmx::evalSolution: Piola mapping requires "
+                 "a single field on each basis" << std::endl;
```
[the remark this answers is not in the document]
Comment (akva2) [reply src/ASM/LR/ASMu2Dmx.C]: ok

File: src/ASM/ASMs2Dmx.C [hunk 7]
```diff
@@ -1027,6 +1027,91 @@ bool ASMs2Dmx::evalSolution (Matrix& sField, const Vector& locSol,
...
+                  [](const unsigned char in) { return in > 1; })) {
+    std::cerr << "*** ASMs2Dmx::evalSolution: Piola mapping requires "
+                 "a single field on each basis" << std::endl;
+    return false;
+  }
+
+  size_t len = std::accumulate(nb.begin(),nb.end(),0u);
+  if (len != locSol.size()) {
+    std::cerr << "*** ASMs2Dmx::evalSolution: Unexpected solution size"
+                 ", expected " << len << ", got " << locSol.size() << std::endl;
```
[the remark this answers is not in the document]
Comment (akva2) [reply src/ASM/ASMs2Dmx.C]: ok

File: src/ASM/ASMs2Dmx.C [hunk 8]
```diff
@@ -1027,6 +1027,91 @@ bool ASMs2Dmx::evalSolution (Matrix& sField, const Vector& locSol,
...
+                                  const RealArray* gpar, bool regular) const
+{
+  // Evaluate the basis functions at all points
+  std::vector<std::vector<Go::BasisPtsSf>> splinex(m_basis.size());
+  std::vector<Go::BasisDerivsSf> splineg;
+
+  if (std::any_of(nfx.begin(), nfx.end(),
+                  [](const unsigned char in) { return in > 1; })) {
+    std::cerr << "*** ASMs2Dmx::evalSolution: Piola mapping requires "
+                 "a single field on each basis" << std::endl;
```
[the remark this answers is not in the document]
Comment (akva2) [reply src/ASM/ASMs2Dmx.C]: while multi-line string concat on new lines is nothing to be surprised by, i have added the stream operator your peace of mind.

File: src/ASM/LR/ASMu3D.h [hunk 9]
```diff
@@ -413,8 +413,9 @@ class ASMu3D : public ASMLRSpline, public ASM3D
   //! \param[in] locSol Solution vector in DOF-order
   //! \param[in] npe Number of visualization nodes over each knot span
   //! \param[in] nf If nonzero, mixed evaluates nf fields on first basis
+  //! \param[in] piola True if solution uses Piola mapping
   virtual bool evalSolution(Matrix& sField, const Vector& locSol,
-                            const int* npe, int nf) const;
+                            const int* npe, int nf, bool piola = false) const;
```
[the remark this answers is not in the document]
Comment (akva2) [reply src/ASM/LR/ASMu3D.h]: ok

File: src/ASM/ASMsupel.h [hunk 10]
```diff
@@ -103,11 +103,12 @@ class ASMsupel : public ASMbase
   //! \param[out] grid The generated finite element grid
   virtual bool tesselate(ElementBlock& grid, const int*) const;
 
+  using ASMbase::evalSolution;
   //! \brief Evaluates the primary solution field at all visualization points.
   //! \param[out] sField Solution field
   //! \param[in] locSol Solution vector local to current patch
   virtual bool evalSolution(Matrix& sField, const Vector& locSol,
-                            const int*, int) const;
+                            const int*, int, bool = false) const;
```
[the remark this answers is not in the document]
Comment (akva2) [reply src/ASM/ASMsupel.h]: ok

File: src/SIM/SIMgeneric.C [hunk 11]
```diff
@@ -53,8 +53,13 @@ Vector SIMgeneric::getSolution (const Vector& psol, const double* par,
...
-    return Vector();
+  if (myProblem->getIntegrandType() & Integrand::PIOLA_MAPPING) {
+    if (!pch->evalSolutionPiola(tmpVal,localVec,&params.front(),false))
+      return Vector();
+  } else {
+    if (!pch->evalSolution(tmpVal,localVec,&params.front(),false,deriv))
+      return Vector();
+  }
 
   return tmpVal.getColumn(1);
```
[the remark this answers is not in the document]
Comment (akva2) [reply src/SIM/SIMgeneric.C]: ok

**Comment by akva2** [comment]
jenkins build this with downstreams ifem-stokes=257 please

**Pull request merged**